apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts

```diff
@@ -0,0 +1,243 @@
+import { loadSqlEquiv, queryRows } from '@prairielearn/postgres';
+
+import { config } from '../../../lib/config.js';
+import { setQuestionCopyTargets } from '../../../lib/copy-question.js';
+import { getCourseFilesClient } from '../../../lib/course-files-api.js';
+import { AiGenerationPromptSchema, IdSchema } from '../../../lib/db-types.js';
+import { features } from '../../../lib/features/index.js';
+import { idsEqual } from '../../../lib/id.js';
+import {
+  getAndRenderVariant,
+  renderPanelsForSubmission,
+  setRendererHeader,
+} from '../../../lib/question-render.js';
+import { processSubmission } from '../../../lib/question-submission.js';
+import { HttpRedirect } from '../../../lib/redirect.js';
+import { logPageView } from '../../../middlewares/logPageView.js';
+import { selectQuestionById } from '../../../models/question.js';
+import { regenerateQuestion } from '../../lib/aiQuestionGeneration.js';
+import { GenerationFailure } from '../instructorAiGenerateQuestion/instructorAiGenerateQuestion.html.js';
+
+import { AiGenerateEditorPage } from './instructorAiGenerateDraftEditor.html.js';
+
+const router = express.Router();
+const sql = loadSqlEquiv(import.meta.url);
+
+export async function saveGeneratedQuestion(
+  res: express.Response,
+  htmlFileContents: string | undefined,
+  pythonFileContents: string | undefined,
+  title?: string,
+  qid?: string,
+): Promise<string> {
+  const files = {};
+
+  if (htmlFileContents) {
+    files['question.html'] = htmlFileContents;
+  }
+
+  if (pythonFileContents) {
+    files['server.py'] = pythonFileContents;
+  }
+
+  const client = getCourseFilesClient();
+
+  const result = await client.createQuestion.mutate({
+    course_id: res.locals.course.id,
+    user_id: res.locals.user.user_id,
+    authn_user_id: res.locals.authn_user.user_id,
+    has_course_permission_edit: res.locals.authz_data.has_course_permission_edit,
+    qid,
+    title,
+    files,
+  });
+
+  if (result.status === 'error') {
+    throw new HttpRedirect(res.locals.urlPrefix + '/edit_error/' + result.job_sequence_id);
+  }
+
+  return result.question_id;
+}
+
+function assertCanCreateQuestion(resLocals: Record<string, any>) {
+  // Do not allow users to edit without permission
+  if (!resLocals.authz_data.has_course_permission_edit) {
+    throw new error.HttpStatusError(403, 'Access denied (must be course editor)');
+  }
+
+  // Do not allow users to edit the exampleCourse
+  if (resLocals.course.example_course) {
+    throw new error.HttpStatusError(403, 'Access denied (cannot edit the example course)');
+  }
+}
+
+router.use(
+  asyncHandler(async (req, res, next) => {
+    if (!(await features.enabledFromLocals('ai-question-generation', res.locals))) {
+      throw new error.HttpStatusError(403, 'Feature not enabled');
+    }
+    next();
+  }),
+);
+
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    assertCanCreateQuestion(res.locals);
+    if (res.locals.question_id) {
```

Comment:
What is expected to set this value? This seems like the sort of thing that should come from `req.params` instead.

Reply:
I changed this to use `req.params` instead in 2695872.